elasticdl/python/master/pod_manager.py

```diff
@@ -418,153 +473,154 @@ def _event_cb(self, event):
+            [
+                callback.on_pod_succeeded(pod_info, cluster_context)
+                for callback in self._pod_event_callbacks
+            ]
+        elif matched_pod_state_flow.to_status == PodStatus.FAILED:
+            [
+                callback.on_pod_failed(pod_info, cluster_context)
+                for callback in self._pod_event_callbacks
+            ]
+            should_relaunch = should_relaunch and _should_relaunch_killed_pod(
+                evt_obj=evt_obj
+            )
+        elif matched_pod_state_flow.to_status == PodStatus.DELETED:
+            [
+                callback.on_pod_deleted(pod_info, cluster_context)
+                for callback in self._pod_event_callbacks
+            ]
+
+        if should_relaunch:
+            logger.info("Relaunch the worker: {}".format(pod_name))
 
-            # Notify each PodEventCallback that PodStarted is fired
-            if evt_type in ["ADDED", "MODIFIED"] and phase == "Running":
-                for callback in self._pod_event_callbacks:
-                    callback.on_pod_started(
-                        PodInfo(type=None, id=None, name=pod_name),
-                        ClusterContext(pod_manager=self),
-                    )
-
-            # Notify each PodEventCallback that PodSucceeded is fired
-            if evt_type == "MODIFIED" and phase == "Succeeded":
-                for callback in self._pod_event_callbacks:
-                    callback.on_pod_succeeded(
-                        PodInfo(type=None, id=None, name=pod_name),
-                        ClusterContext(pod_manager=self),
-                    )
-
-        if relaunch_worker and worker_id >= 0:
-            logger.info("Relaunching worker.")
             new_worker_id = self._next_worker_id_fn()
             with self._lock:
                 self._worker_pod_priority[
                     new_worker_id
-                ] = self._worker_pod_priority[worker_id]
+                ] = self._worker_pod_priority[pod_id]
             self._start_worker(new_worker_id)
-        elif relaunch_ps:
-            logger.info("Relaunching ps.")
-            # Note: the ID and service address for relaunched parameter
-            # server are intentionally left unchanged to support fault
-            # tolerance.
-            self._start_ps(ps_id)
-
-    def check_all_workers_exited(self):
-        workers_exited = []
-        for (_, _, phase,) in self._worker_pods_ip_phase.values():
-            workers_exited.append(
-                phase == PodStatus.FAILED or phase == PodStatus.SUCCEEDED
+
+    @staticmethod
+    def get_pod_state_flow(from_status, event_type, phase):
+        for pod_state_flow in POD_STATE_FLOWS:
+            if (
+                from_status == pod_state_flow.from_status
+                and event_type == pod_state_flow.event_type
+                and (
+                    phase == pod_state_flow.phase
+                    or pod_state_flow.phase is None
```

Comment:
`pod_state_flow.phase is None or phase == pod_state_flow.phase` may be better.

Reply:
Good idea, updated.